payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java

```diff
@@ -462,7 +471,54 @@ public boolean apply(final PaymentTransactionModelDao curPaymentTransactionModel
                                   curPaymentModelDao.getPaymentMethodId(),
                                   curPaymentModelDao.getPaymentNumber(),
                                   curPaymentModelDao.getExternalKey(),
-                                  sortedTransactions);
+                                  sortedTransactions,
+                                  (withAttempts && !sortedTransactions.isEmpty()) ? getNextRetryDateForFailedPayment(sortedTransactions) : null
+                                  );
+    }
+
+    private PaymentAttempt getNextRetryDateForFailedPayment(final List<PaymentTransaction> purchasedTransactions) {
```

Comment:
We should not recompute the payment retry dates but instead fetch the future notifications associated with these pending notifications. You will need to use the notification apis to access that information:
* A code example to fetch such future notification can be found [here](https://github.com/killbill/killbill/blob/master/overdue/src/main/java/org/killbill/billing/overdue/notification/DefaultOverduePosterBase.java#L110)
* For context, those notifications are inserted [here](https://github.com/killbill/killbill/blob/master/payment/src/main/java/org/killbill/billing/payment/retry/BaseRetryService.java#L116). This will give you a clue on which notification `service` and `queueName` to use and what they contain.